### PR TITLE
Added instrumentation of vector TCG operations

### PR DESCRIPTION
**Pull request by damienmaier**
This is the code I added for the support of vector TCG instructions in SymQEMU.

I have documented the functions in the source files. There is also some information about the work done in chapter 6 of my [bachelor thesis](https://dmaier.ch/bachelor-thesis.pdf).

### CODE REVIEW

**Comment by damienmaier** [comment]
Sure !

**Comment by damienmaier** [comment]
Just some infos about how files are organized in this PR :

The original sym helper functions are in `accel/tcg/tcg-runtime-sym.c`.

I put the vector sym helper functions in a new file `accel/tcg/tcg-runtime-sym-vec.c`. I also created `accel/tcg/tcg-runtime-sym-common.c` where I moved some code from `tcg-runtime-sym.c` that is used both for vector and non-vector sym helper functions.

It seemed better to me to have a separate file for vector helper functions, but if your prefer to have everything in `accel/tcg/tcg-runtime-sym.c instead`, I can change that.

Other that that I modified `tcg/tcg-op-vec.c` to inject instrumentation TCG ops. I also added support for helper functions with a larger number of arguments as I needed it.

File: tcg/tcg-op-vec.c [hunk 1]
```diff
@@ -734,16 +986,16 @@ void tcg_gen_bitsel_vec(unsigned vece, TCGv_vec r, TCGv_vec a,
     tcg_debug_assert(bt->base_type >= type);
     tcg_debug_assert(ct->base_type >= type);
 
-    if (TCG_TARGET_HAS_bitsel_vec) {
+    /* if (TCG_TARGET_HAS_bitsel_vec) {
```
Comment (aurelf) [inline tcg/tcg-op-vec.c:989]: Can you explain why this code is commented, and if it needs to be preserved as a comment in the commit? Is it not working on some target arch?

File: tcg/tcg-op-vec.c [hunk 2]
```diff
@@ -772,16 +1024,16 @@ void tcg_gen_cmpsel_vec(TCGCond cond, unsigned vece, TCGv_vec r,
     hold_list = tcg_swap_vecop_list(NULL);
     can = tcg_can_emit_vec_op(INDEX_op_cmpsel_vec, type, vece);
 
-    if (can > 0) {
+    /* if (can > 0) {
```
Comment (aurelf) [inline tcg/tcg-op-vec.c:1027]: Same here and few more below

File: tcg/tcg-op-vec.c [hunk 3]
```diff
@@ -293,76 +428,98 @@ void tcg_gen_stl_vec(TCGv_vec r, TCGv_ptr b, TCGArg o, TCGType low_type)
...
 {
+    vec_vec_op_instrumentation(vece, r, a, b, gen_helper_sym_xor_vec);
     vec_gen_op3(INDEX_op_xor_vec, 0, r, a, b);
 }
 
 void tcg_gen_andc_vec(unsigned vece, TCGv_vec r, TCGv_vec a, TCGv_vec b)
 {
-    if (TCG_TARGET_HAS_andc_vec) {
+   /* if (TCG_TARGET_HAS_andc_vec) {
         vec_gen_op3(INDEX_op_andc_vec, 0, r, a, b);
```
Comment (aurelf) [inline tcg/tcg-op-vec.c:474]: Why is this commented? Is vec_gen_op3 declared above not working (for some operations?) ? 
Same for the 3  others:
INDEX_op_andc_vec
INDEX_op_orc_vec
INDEX_op_nand_vec
INDEX_op_nor_vec
If they need to be implemented ? 
You mention those in section 6.1.7 of the thesis, but I still don't get why the direct support isn't used when the target supports it?

File: tcg/tcg-op-vec.c [hunk 4]
```diff
@@ -393,9 +550,9 @@ void tcg_gen_not_vec(unsigned vece, TCGv_vec r, TCGv_vec a)
 {
     const TCGOpcode *hold_list = tcg_swap_vecop_list(NULL);
 
-    if (!TCG_TARGET_HAS_not_vec || !do_op2(vece, r, a, INDEX_op_not_vec)) {
+    /*if (!TCG_TARGET_HAS_not_vec || !do_op2(vece, r, a, INDEX_op_not_vec)) {*/
```
Comment (aurelf) [inline tcg/tcg-op-vec.c:553]: Same comment as above

File: tcg/tcg-op-vec.c [hunk 5]
```diff
@@ -406,9 +563,9 @@ void tcg_gen_neg_vec(unsigned vece, TCGv_vec r, TCGv_vec a)
     tcg_assert_listed_vecop(INDEX_op_neg_vec);
     hold_list = tcg_swap_vecop_list(NULL);
 
-    if (!TCG_TARGET_HAS_neg_vec || !do_op2(vece, r, a, INDEX_op_neg_vec)) {
+    /*if (!TCG_TARGET_HAS_neg_vec || !do_op2(vece, r, a, INDEX_op_neg_vec)) {*/
```
Comment (aurelf) [inline tcg/tcg-op-vec.c:566]: idem

File: tcg/tcg-op-vec.c [hunk 6]
```diff
@@ -419,7 +576,7 @@ void tcg_gen_abs_vec(unsigned vece, TCGv_vec r, TCGv_vec a)
     tcg_assert_listed_vecop(INDEX_op_abs_vec);
     hold_list = tcg_swap_vecop_list(NULL);
 
-    if (!do_op2(vece, r, a, INDEX_op_abs_vec)) {
+    /*if (!do_op2(vece, r, a, INDEX_op_abs_vec)) {*/
```
Comment (aurelf) [inline tcg/tcg-op-vec.c:579]: idem

**Review by aurelf — COMMENTED** [review]
I added a few comments. It would be nice to clarify those points, if they don't need a change, maybe add a comment on what is needed to to activate those features?

**Comment by aurelf** [comment]
Also it would be nice to merge a PR with your testsuite, before we merge this PR (and you could add the tests that are specific to the vec operations to this PR?)

**Comment by damienmaier** [comment]
Regarding the commented code, the only motivation was to save implementation time. By making those instructions being systematically replaced by equivalent, already instrumented instructions, I avoided the work of instrumenting them.

For my defense, this trick was already used for several instructions in the original SymQEMU :) Here is an example https://github.com/eurecom-s3/symqemu/blob/master/tcg/tcg-op.c#L1059

I left the original code (but commented) to be consistent, because this is how it was done for other non vec instructions, like in example above.

**Comment by damienmaier** [comment]
All right, I will make a PR with the tests. I am working on creating a Dockerfile to compile SymQEMU and run the tests

**Comment by aurelf** [comment]
> Regarding the commented code, the only motivation was to save implementation time. By making those instructions being systematically replaced by equivalent, already instrumented instructions, I avoided the work of instrumenting them.

Ah, OK.  There will be a performance impact I imagine ? The target code could be more efficient if it would be implemented with one instruction ? 

> For my defense, this trick was already used for several instructions in the original SymQEMU :) Here is an example https://github.com/eurecom-s3/symqemu/blob/master/tcg/tcg-op.c#L1059

Fair :) 

> I left the original code (but commented) to be consistent, because this is how it was done for other non vec instructions, like in example above.

OK, if my understanding above is correct: could you add a comment, at least on the first such occurrence. I.e., to explain the instruction instrumentation isn't implemented, the alternative works, implementing the instrumentation may give some performance gains (as some likely more effective instructions will be generated)?

File: accel/tcg/tcg-runtime-sym-vec.c [hunk 7]
```diff
@@ -0,0 +1,516 @@
...
+
+/*
+ * Symbolic equivalent of an SIMD ternary operation.
+ *
+ * This function performs the symbolic equivalent of computing an output vector of the form :
+ * result[i] = arg1[i] <comparison_operator> arg2[i] ? arg1[i] : arg2[i]
+ * where <vector>[i] denotes the ith element of the vector <vector>.
+ *
+ * For each element, the concrete result is used to determine if the ternary condition was true or false,
+ * and path constraints are pushed accordingly.
```
Comment (sebastianpoeplau) [inline accel/tcg/tcg-runtime-sym-vec.c:440]: Same here, let's add the exploration of a symbolic result instead of path constraints as a TODO comment.

**Comment by damienmaier** [comment]
> Ah, OK. There will be a performance impact I imagine ? The target code could be more efficient if it would be implemented with one instruction ?

Yes I think that this has a performance impact.

> OK, if my understanding above is correct: could you add a comment, at least on the first such occurrence. I.e., to explain the instruction instrumentation isn't implemented, the alternative works, implementing the instrumentation may give some performance gains (as some likely more effective instructions will be generated)?

I added TODO comments to clarify the the situation

**Comment by aurelf** [comment]
Thanks for the updates !

**Comment by damienmaier** [comment]
@sebastianpoeplau I added the todos regarding returning expressions instead of pushing path constrains, and I edited the README.

> And do you know whether the test mentioned in the same section still works now that some functions have moved and/or changed signatures

I realize now that the way QEMU tests are run has changed between QEMU 4 and QEMU 8, because QEMU now uses meson. I didn't think about those SymQEMU unit tests when I did the port to QEMU 8, so currently the file `tests/check-sym-runtime.c` is there but I don't know how to run it.

**Comment by aurelf** [comment]
Hi, I rebased this on top of the current main. made a few minor changes, the tests pass. 
I think it's good to merge. 
It would be nice @damienmaier & @sebastianpoeplau if you could have a look and confirm if it is OK for you?

**Comment by damienmaier** [comment]
It is ok for me !

**Comment by aurelf** [comment]
I had to update also the unit tests, which I just re-enabled. 
I'll merge now, after the test passes.
Thanks to both of you!

**Pull request merged**